src/Functions/FunctionsConversion.h

```diff
@@ -263,13 +275,30 @@ struct ConvertImpl
                         }
                         else
                         {
-                            vec_to[i] = static_cast<ToFieldType>(vec_from[i]);
+                            if constexpr (IsDataTypeNumber<FromDataType> && IsDataTypeDateOrDateTime<ToDataType>)
+                            {
+                                if (isLtZero(vec_from[i]))
+                                {
+                                    vec_to[i] = 0;
```

Comment:
For date lower bound is zero, yes, but for date32 negative values are allowed until its offset epoch.
Example: https://github.com/ClickHouse/ClickHouse/blob/b024c215afa89fee914e44fded7000aa27c90995/src/Functions/FunctionsConversion.h#L1236-L1242
i.e. toDateOrNull(-1) should be Null, but toDate32OrNull(-1) is a correct value and should be  1969-12-31.